components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/logout/frontchannel/DynamicLogoutPageBuilderUtil.java

```diff
@@ -0,0 +1,159 @@
+            "<script>\n" +
+                    "\tredirect();\n" +
+                    "\t</script>\n";
+
+    private static final String REDIRECT_URL = "redirectURL";
+    private static final String FRONTCHANNEL_LOGOUT_URL = "frontchannelLogoutURL";
+    private static final String BODY = "body";
+    private static final String FRONTCHANNEL_LOGOUT_URLS_COUNT = "frontchannelLogoutsCount";
+
+    public static String buildPage(HttpServletRequest request) {
+
+        String htmlPage = STATIC_CONTENT_TEMPLATE;
+        StringBuilder body = new StringBuilder();
+
+        List<String> frontchannelLogoutURLs = getFrontchannelLogoutURLs(request);
+
+        if (!frontchannelLogoutURLs.isEmpty()) {
+            htmlPage = replacePlaceholder(htmlPage, FRONTCHANNEL_LOGOUT_URLS_COUNT, String.valueOf(frontchannelLogoutURLs.size()));
+            for (String frontchannelLogoutURL : frontchannelLogoutURLs) {
+                String iframe = replacePlaceholder(IFRAME_TEMPLATE, FRONTCHANNEL_LOGOUT_URL, frontchannelLogoutURL);
+                body.append(iframe);
+            }
+        } else {
+            htmlPage = replacePlaceholder(htmlPage, FRONTCHANNEL_LOGOUT_URLS_COUNT, String.valueOf(0));
+            body.append(REDIRECT_SCRIPT_TEMPLATE);
+        }
+
+        htmlPage = replacePlaceholder(htmlPage, BODY, body.toString());
+        return htmlPage;
+    }
+
+    public static String setRedirectURL(String htmlPage, String redirectURL) {
+
+        if (htmlPage != null) {
+            htmlPage = replacePlaceholder(htmlPage, REDIRECT_URL, redirectURL);
+        }
+        return htmlPage;
+    }
+
+    private static List<String> getFrontchannelLogoutURLs(HttpServletRequest request) {
+
+        List<String> frontchannelLogoutURLs = new ArrayList<>();
+        OIDCSessionState sessionState = OIDCSessionManagementUtil.getSessionState(request);
+        if (sessionState != null) {
+            Set<String> sessionParticipants = OIDCSessionManagementUtil.getSessionParticipants(sessionState);
+            if (!sessionParticipants.isEmpty()) {
+                OAuthAppDO oAuthAppDO;
+
+                for (String clientID : sessionParticipants) {
+                    try {
+                        oAuthAppDO = OIDCSessionManagementUtil.getOAuthAppDO(clientID);
+                        String frontchannelLogoutURL = oAuthAppDO.getFrontchannelLogoutUrl();
+                        if (frontchannelLogoutURL != null) {
+                            if (!frontchannelLogoutURL.equalsIgnoreCase(("null"))) {
+                                frontchannelLogoutURLs.add(frontchannelLogoutURL);
+                            }
+                        }
+                    } catch (IdentityOAuth2Exception e) {
+                        log.error("Error while getting Logout URL for client id: " + clientID, e);
+                    } catch (InvalidOAuthClientException e) {
+                        log.error("Client id " + clientID + "is invalid.", e);
```

Comment:
Would be helpful to add some context to this error log. Maybe mention what flow is this, and the goal we're trying to achieve.